src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts

```diff
@@ -94,59 +106,213 @@ export class InstructorCourseEnrollPageComponent implements OnInit {
+      this.setTableStyleBasedOnFieldChecks(newStudentsHOTInstance, hotInstanceColHeaders);
+      this.isEnrolling = false;
+      return;
+    }
+
+    this.partitionStudentEnrollRequests(Array.from(studentEnrollRequests.values()));
+    const enrolledStudents: Student[] = [];
+
+    // Use concat because we cannot afford to parallelize with forkJoin when there's data dependency
+    const enrollRequests: Observable<Students> = concat(
+        ...this.allStudentChunks.map((studentChunk: StudentEnrollRequest[]) => {
+          const request: StudentsEnrollRequest = {
+            studentEnrollRequests: studentChunk,
+          };
+          return this.studentService.enrollStudents(
+              this.courseId, request,
+          );
+        }),
+    );
+
+    enrollRequests.pipe(finalize(() => this.isEnrolling = false)).subscribe({
+      next: (resp: Students) => {
+        enrolledStudents.push(...resp.students);
+      },
+      complete: () => {
+        this.showEnrollResults = true;
+        this.statusMessage.pop(); // removes any existing error status message
+        this.statusMessageService.showSuccessToast('Enrollment successful. Summary given below.');
+
+        this.enrollResultPanelList = this.populateEnrollResultPanelList(this.existingStudents,
+            enrolledStudents, Array.from(studentEnrollRequests.values()));
+
+        this.studentService.getStudentsFromCourse({ courseId: this.courseId }).subscribe((resp: Students) => {
+          this.existingStudents = resp.students;
+          if (!this.isExistingStudentsPanelCollapsed) {
+            const existingStudentTable: Handsontable = this.hotRegisterer.getInstance(this.existingStudentsHOT);
+            this.loadExistingStudentsData(existingStudentTable, this.existingStudents);
+          }
+          this.isExistingStudentsPresent = true;
+        });
+      },
+      error: (resp: ErrorMessageOutput) => {
+        this.enrollErrorMessage = resp.error.message;
+      },
+    });
+  }
+
+  private partitionStudentEnrollRequests(studentEnrollRequests: StudentEnrollRequest[]): void {
+    let currentStudentChunk: StudentEnrollRequest[] = [];
+    for (const request of studentEnrollRequests) {
+      currentStudentChunk.push(request);
+      if (currentStudentChunk.length >= this.numberOfStudentsPerRequest) {
+        this.allStudentChunks.push(currentStudentChunk);
+        currentStudentChunk = [];
+      }
+    }
+    if (currentStudentChunk.length > 0) {
+      this.allStudentChunks.push(currentStudentChunk);
+    }
+  }
+
+  private checkTeamsValid(studentEnrollRequests: Map<number, StudentEnrollRequest>): void {
+    const teamSectionMap: Map<string, string> = new Map();
+    const teamIndexMap: Map<string, number> = new Map();
+    const noInvalidRowsBefore: number = this.invalidRowsIndex.size;
+
+    Array.from(studentEnrollRequests.keys()).forEach((key: number) => {
+      const request: StudentEnrollRequest | undefined = studentEnrollRequests.get(key);
```

Comment:
when would this be undefined?

Reply:
this would not be undefined, but enforcing non-null is not permitted and so putting `studentEnrollRequests.get(key)!!` will not pass lint